zebra-consensus/src/checkpoint.rs

```diff
@@ -153,169 +263,212 @@ impl CheckpointVerifier {
+        match self.previous_checkpoint_height() {
+            BeforeGenesis => Some(Unbounded),
+            PreviousCheckpoint(height) => Some(Excluded(height)),
+            FinalCheckpoint => None,
         }
     }
 
-    /// Return the next checkpoint height that we want to verify.
+    /// Return the target checkpoint height that we want to verify.
     ///
-    /// The height increases as blocks are verified.
+    /// If we need more blocks, returns `WaitingForBlocks`.
     ///
-    /// If verification has not started yet, returns zero (the genesis block).
-    /// If verification has finished, returns None.
-    fn get_next_checkpoint_height(&self) -> Option<BlockHeight> {
-        match self.current_checkpoint_range {
-            Some((_, Included(height))) => Some(height),
-            None => None,
-            _ => unreachable!(),
+    /// If the queued blocks are continuous from the previous checkpoint to a
+    /// subsequent checkpoint, returns `Checkpoint(height)`. The target
+    /// checkpoint can be multiple checkpoints ahead of the previous checkpoint.
+    /// `height` increases as checkpoints are verified.
+    ///
+    /// If verification has finished, returns `FinishedVerifying`.
+    fn target_checkpoint_height(&self) -> Target<BlockHeight> {
+        // Find the height we want to start searching at
+        let mut pending_height = match self.previous_checkpoint_height() {
+            // Check if we have the genesis block as a special case, to simplify the loop
+            BeforeGenesis if !self.queued.contains_key(&BlockHeight(0)) => return WaitingForBlocks,
+            BeforeGenesis => BlockHeight(0),
+            PreviousCheckpoint(height) => height,
+            FinalCheckpoint => return FinishedVerifying,
+        };
+
+        // Find the end of the continuous sequence of blocks, starting at the
+        // last verified checkpoint. If there is no verified checkpoint, start
+        // *after* the genesis block (which we checked above).
+        //
+        // If `btree_map::Range` implements `ExactSizeIterator`, it would be
+        // much faster to walk the checkpoint list, and compare the length of
+        // the `btree_map::Range` to the block height difference between
+        // checkpoints. (In maps, keys are unique, so we don't need to check
+        // each height value.)
+        //
+        // But at the moment, this implementation is slightly faster, because
+        // it stops after the first gap.
+        for (&height, _) in self.queued.range((Excluded(pending_height), Unbounded)) {
+            // If the queued blocks are continuous.
+            //
+            // Precondition: queued block heights are within the valid
+            // BlockHeight range, so there is no integer overflow.
+            assert!(
+                pending_height.0 < u32::MAX,
+                "pending block had a height of u32::MAX "
+            );
```

Comment:
I don't think this check is required, since we guarantee that a block height is in valid range (`< 500_000_000`) when we parse it (in `parse_coinbase_height`). 
```suggestion
            // If the queued blocks are continuous.
```

Reply:
Thanks, I'd forgotten that detail of `parse_coinbase_height`.

Reply:
Fixed in 94fa441dd1f7961efa4a53f38838ddba21e1dc8a